### ISSUE TITLE
WIKI: What the MIOPEN_FIND_MODE ("FAST" or 2) does.

### ISSUE THREAD

**Issue by sfinktah** [patientx/ComfyUI-Zluda#350]
_in lieu of a wiki, i'm posting wiki-like articles as issues prefaced with `WIKI: `_

_from an excellent post by [Only8Bits](https://github.com/Only8Bits) [here](https://github.com/comfyanonymous/ComfyUI/pull/10302#issuecomment-3425750147)_

Reference: https://rocm.docs.amd.com/projects/MIOpen/en/latest/how-to/find-and-immediate.html

Short version: AMD has 2 card types, pro and consumer. On all cards the default behaviour is to search for optimal compute solution to any given matrix problem the first time it is ecountered - that takes both time and VRAM. After that it'll be stored in local database and looked up so it's fast. The problem is MIOpen database is per-version, that is once the version is bumped the old database is ignored and a new one is started. And each new major ROCm release usually has new MIOpen version. There is a database of optimal compute solutions for pro cards but not consumer ones (maybe it'll change someday but it hasn't for years) so Radeons need to run a lot of computations to figure out what is what - though even for pro cards the AMD database does not have 100% coverage for all possible matrix sizes and data types.

Rather than breaking the whole app you can set MIOPEN_FIND_MODE="FAST" (or 2) on AMD to just skip the solution search on each new problem. On pro cards you won't even feel it much since these are pretty well profiled by AMD, so any long-term performance loss from suboptimal solution picks are minimal. On Radeons that search is quite time consuming (esp. for training and bigger matrix sizes in SDXL, since noboby has the VRAM to run WAN in 720p mode locally on Radeons and these problems tend to be smaller and faster to find) but it usually results in 5% or so reduced compute times long term once an optimal solution is found.

TLDR: Use MIOPEN_FIND_MODE or just accept that (sometimes massive) slowdown on each first time for a few % better performance long term. It's not ComfyUI issue, it's how MIOpen works. And consumer cards suffer the most since there is no global database from AMD, but also tend to find better solutions than the default ones so it's usually worth it.

**Comment by patientx** [comment]
```
set MIOPEN_FIND_MODE=2
set MIOPEN_LOG_LEVEL=3
```

means this.

**Comment by sfinktah** [comment]
And it means setting 
```
set MIOPEN_FIND_MODE=2
```

is the same as
```
set MIOPEN_FIND_MODE=FAST
```

Which was always something that bothered me.

**Comment by PsiloaxinV** [comment]
Old thread, I know. This might be a dumb question, but is MIOpen only available with HIP SDK 6.2.4 (with the ZLUDA HIP SDK extension)?

**Comment by sfinktah** [comment]
It's available with the 6.2 extension, the 6.5 complete (unofficial sdk), therock torch builds, and possibly older 5.x patches.

It usually manifests as `c:\Program Files\AMD\ROCm\6.2\bin\MIOpen.dll` (or what-ever the path is) or somewhere in `venv` if you're using a therock rocm torch wheel.

It's not available as part of any **official** amd sdk release on windows.

**Comment by PsiloaxinV** [comment]
Yeah I figured that was the case. I was just curious why there was no similar extension for 6.4.2.

By the way, I'm currently running the native torch libraries with ROCm 7.1.1, which is the first release for me that's actually outperforming ZLUDA. This one has MIOpen but no AOTriton (disabled for 7000 series cards apparently). Do you know of any custom builds that enable both? I've already set TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1 but I think it's not even compiled with the 7.1.1 release I'm using.

**Comment by sfinktah** [comment]
I have a 7900XTX, and as of last testing (about a month ago), triton (the real one, as used by torch.compile and KJ) is non-functional for that card (possibly other 7 series cards as well).  However, @jammm made a build of @scottt's 6.5.0 native torch wheel that DOES including AOTRITON (no environment variables requires) and supports the 7900XTX/gfx1100.  Details are here:

https://github.com/sfinktah/amd-torch

**Comment by jammm** [comment]
> I have a 7900XTX, and as of last testing (about a month ago), triton (the real one, as used by torch.compile and KJ) is non-functional for that card (possibly other 7 series cards as well). However, [@jammm](https://github.com/jammm) made a build of [@scottt](https://github.com/scottt)'s 6.5.0 native torch wheel that DOES including AOTRITON (no environment variables requires) and supports the 7900XTX/gfx1100. Details are here:
> 
> https://github.com/sfinktah/amd-torch

The latest pytorch 7.1.1 wheels and TheRock nightly wheels do support aotriton now, for both windows and Linux. See https://github.com/ROCm/TheRock/blob/main/RELEASES.md#torch-for-gfx110x-all and https://rocm.docs.amd.com/projects/radeon-ryzen/en/latest/docs/install/installrad/windows/install-pytorch.html

**Comment by PsiloaxinV** [comment]
I installed the latest, but I don't see any difference.

Edit: Nevermind, pytorch-cross-attention actually works now and sampling and vae operations are MUCH faster. However it looks like upscale using model is broken - it just returns static.

Edit2: After one generation everything outputs static now, with encode and sampling steps taking 2-3 seconds. I'll just stick with the 1116 version.

**Comment by patientx** [comment]
OK, lately started getting ridiculous wait times with vae decoding even with therock torch's I am using (it wasn't like this , probably the changes comfyui made recently to the vae decoding "optimizing" made it worse) What took like 1 minute for the first vae decoding now started taking half an hour ... It was like this with zluda but again I stopped having this problem after switching to therock. Whatever the cause it is back, and the solution is keeping the "set MIOPEN_FIND_MODE=2" or "set MIOPEN_FIND_MODE=FAST" PLUS using cudnn toggler before and after vae decoding. Yes, it is like this with zluda and I myself have been using it like that but for now it seems to be the case for newer rocm / rock as well ...

**Comment by sfinktah** [comment]
@patientx that was the general consensus from that whole comfyanonymous breaks comfyui by disabling cudnn, only he didn't do it for rdna2, which really just makes it more confusing for everyone.

suggest just adding `cd custom_plugins; git pull https://github.com/sfinktah/ovum-cudnn-wrapper` -- you are the one that asked for a wrapper that automatically disabled cudnn for vae operations :)
